sql-plugin/src/main/scala/com/nvidia/spark/rapids/AlluxioUtils.scala

```diff
@@ -289,48 +289,90 @@ object AlluxioUtils extends Logging {
     }
 
     if (replaceFunc.isDefined) {
-      def isDynamicPruningFilter(e: Expression): Boolean =
-        e.find(_.isInstanceOf[PlanExpression[_]]).isDefined
-
-      val partitionDirs = relation.location.listFiles(
-        partitionFilters.filterNot(isDynamicPruningFilter), dataFilters)
-
-      // replace all of input files
-      val inputFiles: Seq[Path] = partitionDirs.flatMap(partitionDir => {
-        partitionDir.files.map(f => replaceFunc.get(f.getPath))
-      })
+      def replacePathsInPartitionSpec(spec: PartitionSpec): PartitionSpec = {
+        val partitionsWithPathsReplaced = spec.partitions.map { p =>
+          val replacedPath = replaceFunc.get(p.path)
+          org.apache.spark.sql.execution.datasources.PartitionPath(p.values, replacedPath)
+        }
+        PartitionSpec(spec.partitionColumns, partitionsWithPathsReplaced)
+      }
 
-      // replace all of rootPaths which are already unique
-      val rootPaths = relation.location.rootPaths.map(replaceFunc.get)
+      def createNewFileIndexWithPathsReplaced(
+          spec: PartitionSpec,
+          rootPaths: Seq[Path]): InMemoryFileIndex = {
+        val specAdjusted = replacePathsInPartitionSpec(spec)
+        val replacedPaths = rootPaths.map(p => replaceFunc.get(p))
+        new InMemoryFileIndex(
+          relation.sparkSession,
+          replacedPaths,
+          relation.options,
+          Option(relation.dataSchema),
+          userSpecifiedPartitionSpec = Some(specAdjusted))
+      }
 
-      // check the alluxio paths in root paths exist or not
-      // throw out an exception to stop the job when any of them is not mounted
-      if (replaceMapOption.isDefined) {
-        rootPaths.foreach { rootPath =>
-          replaceMapOption.get.values.find(value => rootPath.toString.startsWith(value)).foreach {
-            matched => checkAlluxioMounted(relation.sparkSession, matched)
+      // If we know the type of file index, try to reuse as much of the existing
+      // FileIndex as we can to keep from having to recalculate it and potentially
+      // mess it up. Things like partitioning are already included and some of the
+      // Spark infer code didn't handle certain types properly. Here we try to just
+      // update the paths to the new Alluxio path. If its not a type of file index
+      // we know then fall back to inferring. The latter happens on certain CSPs
+      // like Databricks where they have customer file index types.
+      if (relation.location.isInstanceOf[PartitioningAwareFileIndex]) {
```

Comment:
it seems we can replace this with a match statement, then `pfi` and `cfi` are obtained from the `case`.